### ISSUE TITLE
Issues mixing Dio with BLOC

### ISSUE THREAD

**Issue by mattRiddoch** [flutterchina/dio#363]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | e.g. flutter / ios / android                          |  flutter  |
| Platform Version               | e.g.  1.5.0 / 12.0 / 9.0                              |   1.5.4-hotfix.2   |
| Dio Version                    | e.g.  2.1.0 / 1.0.17                                  |  dio: ^2.1.7    |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |   VS Code   |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |  100%    |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |   N/A   |
| Demo project link              | e.g. link to a demo project that highlights the issue |   N/A   |

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

Trying to yield BLOC states from inside an interceptor like so:
```
request.interceptors.add(InterceptorsWrapper(onRequest: (RequestOptions options) {
          print('RequestOptions $options');
        }, onError: (DioError error) {
          print('error $error');
          yield LoginState.jwtFailed(error.response.statusCode);//FAILING HERE
        }, onResponse: (Response response) {
          print('response: $response');
        }));
```

I get a 'Error: 'yield' isn't a type.'

I can yield from the catch block like so:

```
on DioError catch (e) {
        // The request was made and the server responded with a status code
        // that falls out of the range of 2xx and is also not 304.
        if (e.response != null) {
         yield LoginState.jwtFailed(e.response.statusCode);
        } 
```

Any tips for yielding a BLOC state from inside the interceptors declaration?

**Comment by amrosama2020** [comment]
I had a similar requirement, what I did is throw an exception and catch this exception in the bloc itself and then yield proper state based on the caught exception.

Honestly I think APIs/repositories shouldn't be concerned with business logic or yield bloc states as it is the responsibility of the bloc.
